File: webapp/src/lib/render/audio-engine.ts
/**
 * Audio Engine for rendering songset audio with gap transitions.
 * 
 * Ports the Python AudioEngine functionality to Node.js using fluent-ffmpeg.
 * Handles audio concatenation, gap calculation (beat-based), and loudness normalization.
 */

import ffmpeg from "fluent-ffmpeg";
import ffmpegStatic from "ffmpeg-static";
import * as fs from "fs/promises";
import * as path from "path";
import { AssetFetcher } from "./asset-fetcher";

if (ffmpegStatic) {
  ffmpeg.setFfmpegPath(ffmpegStatic);
}

export interface SongsetItem {
  id: string;
  songsetId: string;
  songId: string;
  songTitle?: string | null;
  recordingHashPrefix: string | null;
  position: number;
  gapBeats: number | null;
  crossfadeEnabled: number | null;
  crossfadeDurationSeconds: number | null;
  keyShiftSemitones: number | null;
  tempoRatio: number | null;
  tempoBpm?: number | null;
  durationSeconds?: number | null;
}

export interface AudioSegmentInfo {
  item: SongsetItem;
  audioPath: string;
  startTimeSeconds: number;
  durationSeconds: number;
  gapBeforeSeconds: number;
}

export interface ExportResult {
  outputPath: string;
  totalDurationSeconds: number;
  segments: AudioSegmentInfo[];
  sampleRate: number;
  channels: number;
}

export interface AudioEngineOptions {
  targetLufs?: number;
  outputBitrate?: string;
  sampleRate?: number;
  channels?: number;
}

export type ProgressCallback = (currentStep: number, totalSteps: number) => void;

/**
 * AudioEngine handles audio processing for songset rendering.
 * 
 * Features:
 * - Gap transition calculation (beat-based)
 * - Audio concatenation with fluent-ffmpeg
 * - Loudness normalization (RMS-based approximation)
 * - Transition preview generation
 */
export class AudioEngine {
  private assetFetcher: AssetFetcher;
  private targetLufs: number;
  private outputBitrate: string;
  private sampleRate: number;
  private channels: number;
  private previewTempFiles: string[] = [];

  constructor(
    assetFetcher: AssetFetcher,
    options: AudioEngineOptions = {}
  ) {
    this.assetFetcher = assetFetcher;
    this.targetLufs = options.targetLufs ?? -14.0;
    this.outputBitrate = options.outputBitrate ?? "320k";
    this.sampleRate = options.sampleRate ?? 44100;
    this.channels = options.channels ?? 2;
  }

  /**
   * Calculate crossfade overlap in milliseconds.
   */
  private getCrossfadeMs(item: SongsetItem): number {
    if (!item.crossfadeEnabled || !item.crossfadeDurationSeconds) {
      return 0;
    }

    return Math.max(0, Math.round(item.crossfadeDurationSeconds * 1000));
  }

  /**
   * Calculate gap duration in milliseconds based on beats and tempo.
   * 
   * @param item - Songset item with gap configuration
   * @param tempoBpm - Tempo for beat-based gap calculation
   * @returns Gap duration in milliseconds
   */
  calculateGapMs(item: SongsetItem, tempoBpm?: number | null): number {
    // Crossfade - no gap, just the crossfade duration
    if (item.crossfadeEnabled && item.crossfadeDurationSeconds) {
      return 0;
    }

    // Calculate gap from beats (default 2 beats)
    const gapBeats = item.gapBeats ?? 2.0;

    if (tempoBpm && tempoBpm > 0) {
      // Convert beats to milliseconds: 60000ms / BPM = ms per beat
      const beatDurationMs = 60000.0 / tempoBpm;
      return Math.round(gapBeats * beatDurationMs);
    } else {
      // Default: 2 seconds per beat estimate
      return Math.round(gapBeats * 1000);
    }
  }

  /**
   * Get audio file information using ffprobe.
   * 
   * @param filePath - Path to audio file
   * @returns Audio metadata or null if failed
   */
  async getAudioInfo(filePath: string): Promise<{
    durationSeconds: number;
    durationMs: number;
    channels: number;
    sampleRate: number;
    bitrate: number;
    fileSizeBytes: number;
  } | null> {
    try {
      // Check file exists
      const stats = await fs.stat(filePath);
      
      return await new Promise((resolve, reject) => {
        ffmpeg.ffprobe(filePath, (err, metadata) => {
          if (err) {
            reject(err);
            return;
          }

          if (!metadata.streams || metadata.streams.length === 0) {
            reject(new Error("No audio streams found"));
            return;
          }

          const stream = metadata.streams[0];
          const durationSeconds = metadata.format.duration ?? 0;
          const bitrate = parseInt(String(metadata.format.bit_rate ?? "0"), 10);

          resolve({
            durationSeconds,
            durationMs: Math.round(durationSeconds * 1000),
            channels: stream.channels ?? 2,
            sampleRate: stream.sample_rate ?? 44100,
            bitrate: Math.round(bitrate / 1000), // Convert to kbps
            fileSizeBytes: stats.size,
          });
        });
      });
    } catch {
      return null;
    }
  }

  /**
   * Generate combined audio for a songset with gap transitions.
   * 
   * @param items - List of songset items in order
   * @param outputPath - Path for the output audio file
   * @param progressCallback - Called with (currentStep, totalSteps) during processing
   * @param normalize - Whether to normalize loudness
   * @returns ExportResult with output information
   */
  async generateSongsetAudio(
    items: SongsetItem[],
    outputPath: string,
    progressCallback?: ProgressCallback,
    normalize: boolean = true
  ): Promise<ExportResult> {
    if (items.length === 0) {
      throw new Error("Cannot generate audio for empty songset");
    }

    const segments: AudioSegmentInfo[] = [];
    let currentTimeMs = 0;

    const totalSteps = items.length * 2; // Load + process for each item
    let currentStep = 0;

    // Download all audio files first
    const audioFiles: {
      path: string;
      item: SongsetItem;
      gapMs: number;
      crossfadeMs: number;
      durationMs: number;
      startMs: number;
    }[] = [];
    
    for (let i = 0; i < items.length; i++) {
      const item = items[i];
      
      // Update progress
      if (progressCallback) {
        progressCallback(currentStep, totalSteps);
      }
      currentStep++;

      // Download/get cached audio
      if (!item.recordingHashPrefix) {
        throw new Error(`Item ${item.id} has no recording`);
      }

      const audioPath = await this.assetFetcher.downloadAudio(item.recordingHashPrefix);
      if (!audioPath) {
        throw new Error(
          `Could not get audio for recording ${item.recordingHashPrefix}`
        );
      }

      // Calculate gap before this song
      const info = await this.getAudioInfo(audioPath);
      const durationMs = info?.durationMs ?? 0;
      let gapMs = 0;
      let crossfadeMs = 0;

      if (i > 0) {
        gapMs = this.calculateGapMs(item, item.tempoBpm);
        crossfadeMs = Math.min(this.getCrossfadeMs(item), durationMs);
      }

      const startTimeMs = i === 0 ? 0 : Math.max(0, currentTimeMs + gapMs - crossfadeMs);

      // Record segment info
      const segmentInfo: AudioSegmentInfo = {
        item,
        audioPath,
        startTimeSeconds: startTimeMs / 1000.0,
        durationSeconds: durationMs / 1000.0,
        gapBeforeSeconds: gapMs / 1000.0,
      };
      segments.push(segmentInfo);

      currentTimeMs = startTimeMs + durationMs;
      audioFiles.push({
        path: audioPath,
        item,
        gapMs,
        crossfadeMs,
        durationMs,
        startMs: startTimeMs,
      });

      // Update progress
      if (progressCallback) {
        progressCallback(currentStep, totalSteps);
      }
      currentStep++;
    }

    // Ensure output directory exists
    await fs.mkdir(path.dirname(outputPath), { recursive: true });

    // Build FFmpeg command for concatenation
    await this.concatenateAudioFiles(audioFiles, outputPath, normalize);

    // Final progress update
    if (progressCallback) {
      progressCallback(totalSteps, totalSteps);
    }

    return {
      outputPath,
      totalDurationSeconds: currentTimeMs / 1000.0,
      segments,
      sampleRate: this.sampleRate,
      channels: this.channels,
    };
  }

  /**
   * Concatenate audio files with gaps using FFmpeg.
   * 
   * @param audioFiles - Array of audio file paths with gaps
   * @param outputPath - Output file path
   * @param normalize - Whether to apply loudness normalization
   */
  private async concatenateAudioFiles(
    audioFiles: {
      path: string;
      item: SongsetItem;
      gapMs: number;
      crossfadeMs: number;
      durationMs: number;
      startMs: number;
    }[],
    outputPath: string,
    normalize: boolean
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      const command = ffmpeg();
      const filterParts: string[] = [];
      const outputLabels: string[] = [];

      for (const audioFile of audioFiles) {
        command.input(audioFile.path);
      }

      for (let i = 0; i < audioFiles.length; i++) {
        const audioFile = audioFiles[i];
        const nextCrossfadeMs = audioFiles[i + 1]?.crossfadeMs ?? 0;
        const filters = [`[${i}:a]asetpts=PTS-STARTPTS`];

        if (audioFile.crossfadeMs > 0) {
          filters.push(
            `afade=t=in:st=0:d=${(audioFile.crossfadeMs / 1000).toFixed(3)}`
          );
        }

        if (nextCrossfadeMs > 0) {
          const fadeOutStartSeconds = Math.max(
            0,
            (audioFile.durationMs - nextCrossfadeMs) / 1000
          );
          filters.push(
            `afade=t=out:st=${fadeOutStartSeconds.toFixed(3)}:d=${(
              nextCrossfadeMs / 1000
            ).toFixed(3)}`
          );
        }

        if (audioFile.startMs > 0) {
          const delay = Math.round(audioFile.startMs);
          filters.push(`adelay=${delay}|${delay}`);
        }

        const outputLabel = `a${i}`;
        filterParts.push(`${filters.join(",")}[${outputLabel}]`);
        outputLabels.push(`[${outputLabel}]`);
      }

      filterParts.push(
        `${outputLabels.join("")}amix=inputs=${outputLabels.length}:normalize=0:dropout_transition=0[outa]`
      );

      command.complexFilter(filterParts, "[outa]");

      // Apply loudness normalization if requested
      if (normalize) {
        // Use loudnorm filter for EBU R128 loudness normalization
        // target_offset is the target integrated loudness (-14 LUFS is common for streaming)
        command.audioFilters([
          `loudnorm=I=${this.targetLufs}:TP=-1.5:LRA=11`,
        ]);
      }

      command
        .audioCodec("libmp3lame")
        .audioBitrate(this.outputBitrate)
        .audioFrequency(this.sampleRate)
        .audioChannels(this.channels)
        .output(outputPath)
        .on("end", () => resolve())
        .on("error", (err) => reject(err))
        .run();
    });
  }

  /**
   * Generate a preview of a transition between two songs.
   * 
   * @param fromItem - First song item
   * @param toItem - Second song item
   * @param previewDurationSeconds - Duration of the preview clip
   * @returns Path to preview audio file or null if generation failed
   */
  async previewTransition(
    fromItem: SongsetItem,
    toItem: SongsetItem,
    previewDurationSeconds: number = 15.0
  ): Promise<string | null> {
    if (!fromItem.recordingHashPrefix || !toItem.recordingHashPrefix) {
      return null;
    }

    // Clean up previous preview temp files
    await this.cleanupPreviewFiles();

    try {
      // Download audio files
      const fromPath = await this.assetFetcher.downloadAudio(
        fromItem.recordingHashPrefix
      );
      const toPath = await this.assetFetcher.downloadAudio(
        toItem.recordingHashPrefix
      );

      if (!fromPath || !toPath) {
        return null;
      }

      // Get audio info
      const fromInfo = await this.getAudioInfo(fromPath);
      const toInfo = await this.getAudioInfo(toPath);

      if (!fromInfo || !toInfo) {
        return null;
      }

      // Calculate clip durations
      const halfDurationMs = (previewDurationSeconds * 1000) / 2;
      
      // Extract end of first song
      const fromDurationMs = Math.min(halfDurationMs, fromInfo.durationMs);
      const fromStartMs = Math.max(0, fromInfo.durationMs - fromDurationMs);

      // Extract start of second song
      const toDurationMs = Math.min(halfDurationMs, toInfo.durationMs);

      // Calculate gap / overlap
      const gapMs = this.calculateGapMs(toItem, toItem.tempoBpm);
      const crossfadeMs = Math.min(this.getCrossfadeMs(toItem), fromDurationMs, toDurationMs);

      // Create temp file for preview
      const tempFile = path.join(
        await this.assetFetcher.getTempDir(),
        `preview-${Date.now()}.mp3`
      );
      this.previewTempFiles.push(tempFile);

      // Build FFmpeg command for preview
      await new Promise<void>((resolve, reject) => {
        const command = ffmpeg();

        // Add inputs
        command.input(fromPath);
        command.input(toPath);

        // Build filter complex
        const filters: string[] = [];

        // Trim end of first song
        const fromStartSeconds = fromStartMs / 1000;
        const fromDurationSeconds = fromDurationMs / 1000;
        filters.push(
          `[0:a]atrim=start=${fromStartSeconds}:duration=${fromDurationSeconds},asetpts=PTS-STARTPTS[from]`
        );

        // Trim start of second song
        const toDurationSec = toDurationMs / 1000;
        filters.push(
          `[1:a]atrim=start=0:duration=${toDurationSec},asetpts=PTS-STARTPTS[to]`
        );

        if (crossfadeMs > 0) {
          const overlapStartSeconds = Math.max(
            0,
            (fromDurationMs - crossfadeMs) / 1000
          );
          filters.push(
            `[from]afade=t=out:st=${overlapStartSeconds.toFixed(3)}:d=${(
              crossfadeMs / 1000
            ).toFixed(3)}[fromf]`
          );
          filters.push(
            `[to]afade=t=in:st=0:d=${(crossfadeMs / 1000).toFixed(3)},adelay=${Math.max(
              0,
              Math.round(fromDurationMs - crossfadeMs)
            )}|${Math.max(0, Math.round(fromDurationMs - crossfadeMs))}[tof]`
          );
          filters.push("[fromf][tof]amix=inputs=2:normalize=0:dropout_transition=0[outa]");
        } else if (gapMs > 0) {
          const gapSeconds = gapMs / 1000;
          filters.push(
            `aevalsrc=0:d=${gapSeconds}[gap]`
          );
          filters.push("[from][gap][to]concat=n=3:v=0:a=1[outa]");
        } else {
          filters.push("[from][to]concat=n=2:v=0:a=1[outa]");
        }

        command.complexFilter(filters, "[outa]");

        command
          .audioCodec("libmp3lame")
          .audioBitrate("192k")
          .audioFrequency(this.sampleRate)
          .audioChannels(this.channels)
          .output(tempFile)
          .on("end", () => resolve())
          .on("error", (err) => reject(err))
          .run();
      });

      return tempFile;
    } catch {
      return null;
    }
  }

  /**
   * Clean up temporary preview files.
   */
  async cleanupPreviewFiles(): Promise<void> {
    for (const file of this.previewTempFiles) {
      try {
        await fs.unlink(file);
      } catch {
        // Ignore errors
      }
    }
    this.previewTempFiles = [];
  }

  /**
   * Calculate the total duration of a songset including gaps.
   * 
   * @param items - List of songset items
   * @returns Total duration in seconds
   */
  async calculateTotalDuration(items: SongsetItem[]): Promise<number> {
    let totalMs = 0;

    for (let i = 0; i < items.length; i++) {
      const item = items[i];

      // Add gap (except for first song)
      if (i > 0) {
        const gapMs = this.calculateGapMs(item, item.tempoBpm);
        totalMs += gapMs;
        totalMs -= this.getCrossfadeMs(item);
      }

      // Add song duration if available
      if (item.durationSeconds) {
        totalMs += item.durationSeconds * 1000;
      } else if (item.recordingHashPrefix) {
        // Try to get from audio file
        const audioPath = await this.assetFetcher.downloadAudio(
          item.recordingHashPrefix
        );
        if (audioPath) {
          const info = await this.getAudioInfo(audioPath);
          if (info) {
            totalMs += info.durationMs;
          }
        }
      }
    }

    return totalMs / 1000.0;
  }
}

File: services/render-worker/src/sow_render_worker/audio_engine.py
from __future__ import annotations

import json
import logging
import subprocess
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, Protocol

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class SongsetItem:
    id: str
    songset_id: str
    song_id: str
    song_title: str | None = None
    recording_hash_prefix: str | None = None
    position: int = 0
    gap_beats: float | None = None
    crossfade_enabled: int | None = None
    crossfade_duration_seconds: float | None = None
    key_shift_semitones: float | None = None
    tempo_ratio: float | None = None
    tempo_bpm: float | None = None
    duration_seconds: float | None = None


@dataclass(frozen=True)
class AudioSegmentInfo:
    item: SongsetItem
    audio_path: str
    start_time_seconds: float
    duration_seconds: float
    gap_before_seconds: float


@dataclass(frozen=True)
class ExportResult:
    output_path: str
    total_duration_seconds: float
    segments: tuple[AudioSegmentInfo, ...] = field(default_factory=tuple)
    sample_rate: int = 44100
    channels: int = 2


class AssetFetcherProtocol(Protocol):
    def download_audio(self, hash_prefix: str) -> str | None: ...


def get_crossfade_ms(item: SongsetItem) -> int:
    if not item.crossfade_enabled or not item.crossfade_duration_seconds:
        return 0
    return max(0, round(item.crossfade_duration_seconds * 1000))


def calculate_gap_ms(item: SongsetItem, tempo_bpm: float | None = None) -> int:
    if item.crossfade_enabled and item.crossfade_duration_seconds:
        return 0
    gap_beats = item.gap_beats if item.gap_beats is not None else 2.0
    if tempo_bpm and tempo_bpm > 0:
        beat_duration_ms = 60000.0 / tempo_bpm
        return round(gap_beats * beat_duration_ms)
    else:
        return round(gap_beats * 1000)


def get_audio_info(file_path: str) -> dict[str, Any] | None:
    try:
        p = Path(file_path)
        if not p.exists():
            return None
        result = subprocess.run(
            [
                "ffprobe",
                "-v",
                "quiet",
                "-print_format",
                "json",
                "-show_format",
                "-show_streams",
                file_path,
            ],
            capture_output=True,
            text=True,
            timeout=30,
        )
        if result.returncode != 0:
            return None
        metadata = json.loads(result.stdout)
        streams = metadata.get("streams", [])
        if not streams:
            return None
        audio_stream = None
        for s in streams:
            if s.get("codec_type") == "audio":
                audio_stream = s
                break
        if audio_stream is None:
            audio_stream = streams[0]
        fmt = metadata.get("format", {})
        duration_seconds = float(fmt.get("duration", 0))
        bitrate = int(fmt.get("bit_rate", "0") or "0")
        return {
            "duration_seconds": duration_seconds,
            "duration_ms": round(duration_seconds * 1000),
            "channels": int(audio_stream.get("channels", 2)),
            "sample_rate": int(audio_stream.get("sample_rate", 44100)),
            "bitrate_kbps": round(bitrate / 1000),
            "file_size_bytes": p.stat().st_size,
        }
    except Exception:
        return None


def build_ffmpeg_filter_complex(
    audio_files: list[dict[str, Any]],
    normalize: bool,
    target_lufs: float = -14.0,
) -> str:
    filter_parts: list[str] = []
    output_labels: list[str] = []

    for i, audio_file in enumerate(audio_files):
        next_crossfade_ms = audio_files[i + 1]["crossfade_ms"] if i + 1 < len(audio_files) else 0
        filters = [f"[{i}:a]asetpts=PTS-STARTPTS"]

        if audio_file["crossfade_ms"] > 0:
            fade_in_dur = audio_file["crossfade_ms"] / 1000
            filters.append(f"afade=t=in:st=0:d={fade_in_dur:.3f}")

        if next_crossfade_ms > 0:
            fade_out_start = max(0, (audio_file["duration_ms"] - next_crossfade_ms) / 1000)
            fade_out_dur = next_crossfade_ms / 1000
            filters.append(f"afade=t=out:st={fade_out_start:.3f}:d={fade_out_dur:.3f}")

        if audio_file["start_ms"] > 0:
            delay = round(audio_file["start_ms"])
            filters.append(f"adelay={delay}|{delay}")

        output_label = f"a{i}"
        filter_parts.append(f"{','.join(filters)}[{output_label}]")
        output_labels.append(f"[{output_label}]")

    amix_out_label = "[amix_out]" if normalize else "[outa]"
    filter_parts.append(
        f"{''.join(output_labels)}amix=inputs={len(output_labels)}"
        f":normalize=0:dropout_transition=0{amix_out_label}"
    )

    if normalize:
        filter_parts.append(
            f"{amix_out_label}loudnorm=I={target_lufs}:TP=-1.5:LRA=11[outa]"
        )

    return ";".join(filter_parts)


def concatenate_audio_files(
    audio_files: list[dict[str, Any]],
    output_path: str,
    normalize: bool = True,
    target_lufs: float = -14.0,
    output_bitrate: str = "320k",
    sample_rate: int = 44100,
    channels: int = 2,
    job_id: str | None = None,
) -> None:
    filter_complex = build_ffmpeg_filter_complex(audio_files, normalize, target_lufs)

    cmd: list[str] = ["ffmpeg", "-y"]
    for audio_file in audio_files:
        cmd.extend(["-i", audio_file["path"]])
    cmd.extend(["-filter_complex", filter_complex, "-map", "[outa]"])
    cmd.extend(["-c:a", "libmp3lame", "-b:a", output_bitrate])
    cmd.extend(["-ar", str(sample_rate), "-ac", str(channels)])
    cmd.append(output_path)

    logger.info("[%s] FFmpeg audio concat: starting (timeout=1800s)", job_id or "unknown")
    subprocess.run(cmd, check=True, capture_output=True, timeout=1800)
    logger.info("[%s] FFmpeg audio concat: complete", job_id or "unknown")


def generate_songset_audio(
    items: list[SongsetItem],
    output_path: str,
    asset_fetcher: AssetFetcherProtocol,
    progress_callback: Callable[[int, int], None] | None = None,
    normalize: bool = True,
    target_lufs: float = -14.0,
    output_bitrate: str = "320k",
    sample_rate: int = 44100,
    channels: int = 2,
    job_id: str | None = None,
) -> ExportResult:
    if not items:
        raise ValueError("Cannot generate audio for empty songset")

    segments: list[AudioSegmentInfo] = []
    current_time_ms = 0
    total_steps = len(items) * 2
    current_step = 0

    audio_files: list[dict[str, Any]] = []

    for i, item in enumerate(items):
        if progress_callback:
            progress_callback(current_step, total_steps)
        current_step += 1

        if not item.recording_hash_prefix:
            raise ValueError(f"Item {item.id} has no recording")

        logger.info(
            "[%s] Audio: processing song %d/%d - %s (hash=%s)",
            job_id or "unknown", i + 1, len(items),
            item.song_title or "untitled", item.recording_hash_prefix or "N/A",
        )

        audio_path = asset_fetcher.download_audio(item.recording_hash_prefix)
        if not audio_path:
            raise ValueError(f"Could not get audio for recording {item.recording_hash_prefix}")

        info = get_audio_info(audio_path)
        if not info:
            raise ValueError(f"Could not probe audio file: {audio_path}")
        duration_ms = info["duration_ms"]

        gap_ms = 0
        crossfade_ms = 0
        if i > 0:
            gap_ms = calculate_gap_ms(item, item.tempo_bpm)
            crossfade_ms = min(get_crossfade_ms(item), duration_ms)

        logger.info(
            "[%s] Audio: song %d probed - duration=%.1fs, gap_ms=%d, crossfade_ms=%d",
            job_id or "unknown", i + 1, duration_ms / 1000.0, gap_ms, crossfade_ms,
        )

        start_time_ms = 0 if i == 0 else max(0, current_time_ms + gap_ms - crossfade_ms)

        segment_info = AudioSegmentInfo(
            item=item,
            audio_path=audio_path,
            start_time_seconds=start_time_ms / 1000.0,
            duration_seconds=duration_ms / 1000.0,
            gap_before_seconds=gap_ms / 1000.0,
        )
        segments.append(segment_info)

        current_time_ms = start_time_ms + duration_ms
        audio_files.append(
            {
                "path": audio_path,
                "item": item,
                "gap_ms": gap_ms,
                "crossfade_ms": crossfade_ms,
                "duration_ms": duration_ms,
                "start_ms": start_time_ms,
            }
        )

        if progress_callback:
            progress_callback(current_step, total_steps)
        current_step += 1

    output_dir = Path(output_path).parent
    output_dir.mkdir(parents=True, exist_ok=True)

    logger.info(
        "[%s] Audio: starting FFmpeg concatenation of %d files -> %s",
        job_id or "unknown", len(audio_files), output_path,
    )

    concatenate_audio_files(
        audio_files,
        output_path,
        normalize=normalize,
        target_lufs=target_lufs,
        output_bitrate=output_bitrate,
        sample_rate=sample_rate,
        channels=channels,
        job_id=job_id,
    )

    logger.info(
        "[%s] Audio: concatenation complete, total duration=%.1fs, %d segments",
        job_id or "unknown", current_time_ms / 1000.0, len(segments),
    )

    if progress_callback:
        progress_callback(total_steps, total_steps)

    return ExportResult(
        output_path=output_path,
        total_duration_seconds=current_time_ms / 1000.0,
        segments=tuple(segments),
        sample_rate=sample_rate,
        channels=channels,
    )


def calculate_total_duration(
    items: list[SongsetItem],
    asset_fetcher: AssetFetcherProtocol,
) -> float:
    total_ms = 0

    for i, item in enumerate(items):
        if i > 0:
            gap_ms = calculate_gap_ms(item, item.tempo_bpm)
            total_ms += gap_ms
            total_ms -= get_crossfade_ms(item)

        if item.duration_seconds is not None:
            total_ms += item.duration_seconds * 1000
        elif item.recording_hash_prefix:
            audio_path = asset_fetcher.download_audio(item.recording_hash_prefix)
            if audio_path:
                info = get_audio_info(audio_path)
                if info:
                    total_ms += info["duration_ms"]

    return total_ms / 1000.0
